homeassistant/components/nest/climate_sdm.py

```diff
@@ -0,0 +1,331 @@
+        if trait_name not in self._device.traits:
+            return None
+        return self._device.traits[trait_name]
+
+    @property
+    def hvac_mode(self):
+        """Return the current operation (e.g. heat, cool, idle)."""
+        if ThermostatEcoTrait.NAME in self._device.traits:
+            trait = self._device.traits[ThermostatEcoTrait.NAME]
+            if trait.mode == THERMOSTAT_ECO_MODE:
+                return HVAC_MODE_AUTO
+        if ThermostatModeTrait.NAME in self._device.traits:
+            trait = self._device.traits[ThermostatModeTrait.NAME]
+            if trait.mode in THERMOSTAT_MODE_MAP:
+                return THERMOSTAT_MODE_MAP[trait.mode]
+        return HVAC_MODE_OFF
+
+    @property
+    def hvac_modes(self):
+        """List of available operation modes."""
+        supported_modes = []
+        if ThermostatModeTrait.NAME in self._device.traits:
+            trait = self._device.traits[ThermostatModeTrait.NAME]
+            for mode in trait.available_modes:
+                if mode in THERMOSTAT_MODE_MAP:
+                    supported_modes.append(THERMOSTAT_MODE_MAP[mode])
+        return supported_modes
+
+    @property
+    def hvac_action(self):
+        """Return the current HVAC action (heating, cooling)."""
+        if ThermostatHvacTrait.NAME not in self._device.traits:
+            return None
+        trait = self._device.traits[ThermostatHvacTrait.NAME]
+        if trait.status in THERMOSTAT_HVAC_STATUS_MAP:
+            return THERMOSTAT_HVAC_STATUS_MAP[trait.status]
+        return None
+
+    @property
+    def preset_mode(self):
+        """Return the current active preset."""
+        if ThermostatEcoTrait.NAME in self._device.traits:
+            trait = self._device.traits[ThermostatEcoTrait.NAME]
+            return PRESET_MODE_MAP.get(trait.mode, PRESET_NONE)
+        return PRESET_NONE
+
+    @property
+    def preset_modes(self):
+        """Return the available presets."""
+        modes = []
+        if ThermostatEcoTrait.NAME in self._device.traits:
+            trait = self._device.traits[ThermostatEcoTrait.NAME]
+            for mode in trait.available_modes:
+                if mode in PRESET_MODE_MAP:
+                    modes.append(PRESET_MODE_MAP[mode])
+        return modes
+
+    @property
+    def fan_mode(self):
+        """Return the current fan mode."""
+        if FanTrait.NAME in self._device.traits:
+            trait = self._device.traits[FanTrait.NAME]
+            return FAN_MODE_MAP.get(trait.timer_mode, FAN_OFF)
+        return FAN_OFF
+
+    @property
+    def fan_modes(self):
+        """Return the list of available fan modes."""
+        if FanTrait.NAME in self._device.traits:
+            return list(FAN_INV_MODE_MAP.keys())
+        return []
+
+    @property
+    def supported_features(self):
+        """Bitmap of supported features."""
+        features = 0
+        if ThermostatTemperatureSetpointTrait.NAME in self._device.traits:
```

Comment:
Is this something we could calculate inside `async_added_to_hass` ?

Reply:
I am not totally sure I am implementing this correctly, and would like some guidance here.  In particular, the supported mode for target temperature vs target temperature range is based on the current HVAC mode, which is something that will change on update.

I *think* either I am implementing this wrong, and it should support both modes, or it needs to be refreshed any time there is a trait update to check the new mode.   I am not sure i trust the other implementations.

Is it better to always support a range and have a None high/low temp? or only support that feature when in the right mode?   I'd like to write a ClimateEntity spec / documentation after this.

Reply:
I looked at the zha climate implementation and modeled behavior based on that.

Reply:
Yeah sorry, it's a mess. This could use some improvements.